perf/impl/go-libp2p/v0.27/perf.go

```diff
@@ -0,0 +1,126 @@
+package main
+
+import (
+	"context"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"time"
+
+	logging "github.com/ipfs/go-log/v2"
+	pool "github.com/libp2p/go-buffer-pool"
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/network"
+	"github.com/libp2p/go-libp2p/core/peer"
+)
+
+var log = logging.Logger("perf")
+
+const (
+	BlockSize = 64 << 10
+
+	ID = "/perf/1.0.0"
+)
+
+type PerfService struct {
+	Host host.Host
+}
+
+func NewPerfService(h host.Host) *PerfService {
+	ps := &PerfService{h}
+	h.SetStreamHandler(ID, ps.PerfHandler)
+	return ps
+}
+
+func (p *PerfService) PerfHandler(s network.Stream) {
+	u64Buf := make([]byte, 8)
+	_, err := io.ReadFull(s, u64Buf)
+	if err != nil {
+		log.Errorw("err", err)
+		s.Reset()
+		return
+	}
+
+	bytesToSend := binary.BigEndian.Uint64(u64Buf)
+
+	_, err = p.drainStream(context.Background(), s)
+	if err != nil {
+		log.Errorw("err", err)
+		s.Reset()
+		return
+	}
+
+	err = p.sendBytes(context.Background(), s, bytesToSend)
+	if err != nil {
+		log.Errorw("err", err)
+		s.Reset()
+		return
+	}
+
+}
+
+func (ps *PerfService) sendBytes(ctx context.Context, s network.Stream, bytesToSend uint64) error {
+	buf := pool.Get(BlockSize)
+	defer pool.Put(buf)
+
+	for bytesToSend > 0 {
+		toSend := buf
+		if bytesToSend < BlockSize {
+			toSend = buf[:bytesToSend]
+		}
+
+		n, err := s.Write(toSend)
+		if err != nil {
+			return err
+		}
+		bytesToSend -= uint64(n)
+	}
+	s.CloseWrite()
+
+	return nil
+}
+
+func (ps *PerfService) drainStream(ctx context.Context, s network.Stream) (uint64, error) {
+	var recvd int64
+	recvd, err := io.Copy(io.Discard, s)
```

Comment:
nit: Using `io.CopyBuffer` with a 64k buffer might be nicer than allocating one everytime.